Add handle property to flow extension specs

nsimone opened the pull request:
### WHY are these changes introduced?

Partial https://github.com/Shopify/flow/issues/19492

### WHAT is this pull request doing?

Adds `handle` to the flow_action and flow_trigger specs and templates.

### How to test your changes?

`pnpm shopify app generate extension --path <PATH>`
select `Flow action` or `Flow trigger`

### Measuring impact

How do we know this change was effective? Please choose one:

- [X] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [X] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [X] I've considered possible [documentation](https://shopify.dev) changes
- [X] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

omnisyle commented on packages/app/templates/extensions/projects/flow_trigger/shopify.extension.toml.liquid:
```diff
@@ -1,4 +1,5 @@
 name = "{{ name }}"
+handle = "your-handle"
```
> If a user creates 2 actions without modifying this value, it should fail right?

there's a discussion going on in https://github.com/Shopify/shopify/pull/435644 about the validations for `handle` property. I think generating the extension in the CLI won't fail when the `handle` are the same but it'll fail when dev tries to deploy/app dev

nsimone replied:
Updated both the template and the import to use the folder name

omnisyle commented on packages/app/src/cli/services/flow/extension-to-toml.ts:
```diff
@@ -33,6 +33,7 @@ export function buildTomlObject(extension: ExtensionRegistration) {
 
   const localExtensionRepresentation = {
     name: extension.title,
+    handle: extension.uuid,
```
@MaxLeduc do we want to make this the folder name too?

isaacroldan replied:
I think we should, the folder name is already sanitized so it makes sense

MaxLeduc commented on packages/app/src/cli/models/extensions/specifications/flow_action.ts:
```diff
@@ -12,6 +13,7 @@ import {zod} from '@shopify/cli-kit/node/schema'
 
 export const FlowActionExtensionSchema = BaseSchema.extend({
   name: zod.string(),
+  handle: zod.string().refine(validateHandle),
```
Shouldn't the handle be on the `extension` directly? We can double check with @isaacroldan but I believe that's where the CLI is heading since there could be more than one extension per file.

nsimone replied:
I believe you're right, will update. The specification type is in both places so I wasn't sure, but the prototype code has it inside the extensions array.

MaxLeduc reviewed (commented):
Left a couple comments, LMK what you think!

nsimone commented on packages/app/src/cli/services/flow/extension-to-toml.ts:
```diff
@@ -33,6 +34,7 @@ export function buildTomlObject(extension: ExtensionRegistration) {
 
   const localExtensionRepresentation = {
     name: config.title,
+    handle: slugify(config.title ?? ''),
```
Ah perfect, will remove from this PR

nsimone commented on packages/app/src/cli/services/flow/validation.ts:
```diff
@@ -69,3 +69,47 @@ export const validateCustomConfigurationPageConfig = (
+      {
+        code: zod.ZodIssueCode.custom,
+        path: ['handle'],
+        message: "Handle can't exceed 100 characters",
```
These validations are from the discussion on this PR: https://github.com/Shopify/shopify/pull/435644#issuecomment-1625635616

MaxLeduc replied:
Reading that comment we might be missing allowing forward slash - do we want to include that as well?

MaxLeduc replied:
Updated comment [here](https://github.com/Shopify/shopify/pull/435644#issuecomment-1629915032) - does that apply to the handle?

isaacroldan replied:
all these errors can be defined in the schema directly:

```
  handle: zod
    .string()
    .trim()
    .min(3, 'Handle needs to have at least 3 characters')
    .max(100, "Handle can't exceed 100 characters")
    .regex(/^[a-zA-Z0-9.\-_]*$/, 'Handle can only contain alphanumeric characters, periods, hyphens, and underscores'),
```

nsimone commented on packages/app/src/cli/services/generate/extension.ts:
```diff
@@ -182,6 +183,7 @@ async function uiExtensionInit({directory, url, app, name, extensionFlavor}: Ext
           await recursiveLiquidTemplateCopy(templateDirectory, directory, {
             srcFileExtension,
             name,
+            handle: slugify(name),
```
Correct, it's what allows us to use it here: https://github.com/Shopify/cli/pull/2366/files#diff-7d9aa32e49142b0d5c9908c8113093f5c47b12df6c2cde80dfe262874b8d8ac7R2

MaxLeduc commented on packages/app/src/cli/services/flow/validation.test.ts:
```diff
@@ -165,3 +165,73 @@ describe('validateCustomConfigurationPageConfig', () => {
+  test('should throw an error if handle is too long', () => {
+    expect(() =>
+      validateHandle(
+        'some-really-long-handle-over-100-characters-aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa',
```
Love this 😆

nsimone commented:
Rebased on top of https://github.com/Shopify/cli/pull/2352 and removed the handle validation - this is now being taken care of at the base schema level.
